services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,488 @@
+        }
+    }
+
+    /**
+     * This endpoint is used by end-users to start the flow to authorize a device.
+     *
+     * @param userCode the user code to authorize
+     * @return
+     */
+    @GET
+    public Response verifyUserCode(@QueryParam("user_code") String userCode) {
+        event.event(EventType.OAUTH2_DEVICE_VERIFY_USER_CODE);
+
+        checkSsl();
+        checkRealm();
+
+        // So back button doesn't work
+        CacheControlUtil.noBackButtonCacheControlHeader();
+
+        AuthenticationSessionModel authenticationSession = null;
+
+        // try to recover the authentication session in case the user is pushing the code
+        if (HttpMethod.POST.equalsIgnoreCase(httpRequest.getHttpMethod())) {
+            String code = httpRequest.getDecodedFormParameters().getFirst(SESSION_CODE);
+            String tabId = session.getContext().getUri().getQueryParameters().getFirst(Constants.TAB_ID);
+
+            String systemClientId = SystemClientUtil.getSystemClient(realm).getClientId();
+
+            SessionCodeChecks checks = checksForCode(null, code, null, systemClientId, tabId, OAUTH2_USER_CODE_VERIFY);
+
+            if (!checks.verifyActiveAndValidAction(AuthenticationSessionModel.Action.USER_CODE_VERIFICATION.name(),
+                ClientSessionCode.ActionType.LOGIN)) {
+                return checks.getResponse();
+            }
+
+            authenticationSession = checks.getAuthenticationSession();
+        }
+
+        // code is not known, we can infer the client neither. ask the user to provide the code.
+        if (StringUtil.isNullOrEmpty(userCode)) {
+            ClientModel client = SystemClientUtil.getSystemClient(realm);
+
+            if (authenticationSession == null) {
+                authenticationSession = createAuthenticationSession(client);
+            }
+
+            return createVerificationPage(null, authenticationSession);
+        } else {
+            // code exists, probably due to using a verification_uri_complete. Start the authentication considering the client
+            // that started the flow.
+            OAuth2DeviceTokenStoreProvider store = session.getProvider(OAuth2DeviceTokenStoreProvider.class);
+            OAuth2DeviceUserCodeProvider userCodeProvider = session.getProvider(OAuth2DeviceUserCodeProvider.class);
+            String formattedUserCode = userCodeProvider.format(userCode);
+            OAuth2DeviceCodeModel deviceCode = store.getByUserCode(realm, formattedUserCode);
+
+            if (deviceCode == null) {
+                event.error(Errors.INVALID_OAUTH2_USER_CODE);
+                return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+            }
+
+            ClientModel client = realm.getClientByClientId(deviceCode.getClientId());
+
+            authenticationSession = createAuthenticationSession(client);
```

Comment:
This is wrong as it will sometimes creates the duplicated authenticationSession. For the case when you don't use `verification_uri_complete` and the screen is confirmed, this method is called from the `verifyUserCode` POST method. That checks the existing authenticationSession (lines 184-197), but then also creates the new authenticationSession here.

If it is done the approach with `isActionRequest` parameter, you will create authenticationSession here just if this is not actionRequest. For the actionRequest, there should be a hard requirement that authenticationSession already exists (as done in the block 184-197).

Reply:
Actually, we should avoid creating authentication sessions when doing GET. But only when doing POST + valid `user_code`.

The verification page should be completely stateless where an authentication session is only needed when redirecting the user to authenticate and confirm the consent.

Reply:
@pedroigor +1 . Considering that we will always show consent page, it is fine to postpone creation of authenticationSession to the latest possible stage.

Reply:
I fixed.